### PR TITLE
Refactor new tokenizer

### PR DESCRIPTION
**Pull request by mattwynne**
Gets rid of almost all the mutable state by using a `Cursor` to move along the input string.

The `Cursor#scan` method emits a token and returns a new `Cursor` positioned after the token that was emitted.

### CODE REVIEW

File: javascript/test/tokenizeTest.ts [hunk 1]
```diff
@@ -1,121 +1,137 @@
...
-    if (tokenType !== TokenType.text) {
-      tokens.push(new Token(tokenType, input[currentIndex], currentIndex, currentIndex + 1))
+  scan(emit: (token: Token) => void): Cursor {
+    if (this.atStartOfWord) {
+      const word = this.input.slice(this.currentIndex, this.endOfCurrentWord)
+      emit(new Token(TokenType.text, word, this.currentIndex, this.endOfCurrentWord))
+      return new Cursor(this.input, this.endOfCurrentWord)
     }
+
+    emit(
```
Comment (mpkorstanje) [inline javascript/test/tokenizeTest.ts:49]: Not every language supports generators. We have to implement this in Go and Java too. Which will be easier if we use a smaller subset of language features.

Additionally this creates quite a bit of pressure on the garbage collector by creating disposable objects in a loop.

**Comment by mpkorstanje** [comment]
@mattwynne I get the impression you're trying to implement a generator or iterator. You could use `yield` but bear in mind that this construct is not available in all languages. So might be better to stick with the loop version first and extract the iterator after implementing the other languages.

**Comment by mattwynne** [comment]
The idea I wanted to get out of my head was more about using State or Strategy to model the different modes/states we talked about - gobbling up a word, or nibbling away at single characters - but in the end the code didn't go that way, and both those strategies are in the `Cursor#scan` method. 🤷

The iterator thing wasn't really the aim, just how the code ended up. There's a slightly earlier version [here](https://github.com/cucumber/cucumber-expressions/blob/5695208bd16f60d52f964362deaa00781d61e296/javascript/test/tokenizeTest.ts#L82) that still uses an integer index, and just leans on the cursor to help make sense of the input string at/around the current code point.

I hadn't really considered the point about making the code portable, thanks! I should have been clearer that I don't share this as "I think this is the right way to factor the code" so much as just some ideas / inspiration for when we work on the code together next.

**Comment by aslakhellesoy** [comment]
Hi @mattwynne do you intend to continue on this work? I'm a little hazy on the benefits of this.

**Comment by mattwynne** [comment]
@aslakhellesoy 🤷 - I think the question is whether we want to continue on the [new-tokenizer](https://github.com/cucumber/cucumber-expressions/tree/new-tokenizer) branch at all at the moment. I think this has dropped down our priority list and could be abandoned.

**Comment by aslakhellesoy** [comment]
I hadn't noticed there is a `new-tokenizer` branch. Is there an associated issue or PR for that?

**Comment by mattwynne** [comment]
No, it never got far enough to be worth a PR I don't think. It's basically all still in @mpkorstanje's head.

**Pull request closed without merging**